### PR TITLE
Fix DSv4 metadata chunking for mixed long prefill

### PR DESCRIPTION
**Pull request by foraxe**
## Summary

This PR fixes the DSv4 GB200 mixed prefill/decode metadata failure by chunking compressed-attention metadata work at real request/prefill segment boundaries.

The earlier operational workaround was to launch with `--max-prefill-tokens 8192`, which prevented the scheduler from admitting multiple 4k prefills in the same step. This PR does not throttle scheduling. Instead, it lets the scheduler produce the mixed shapes and splits the FlashMLA / DeepGEMM paged-MQA metadata work into legal request-boundary chunks derived from `extend_seq_lens_cpu`.

Key behavior:
- builds DSv4 metadata chunk ranges from scheduler request segments, not arbitrary row counts
- keeps the single full-range fast path native
- creates fresh FlashMLA scheduler metadata per chunk
- applies the same request-boundary chunk plan to DeepGEMM/C4 paged-MQA metadata/logits work
- strips the private `_chunk_ranges` key before native FlashMLA calls, including decode/no-chunk paths

## Why

On GB200, DeepSeek-V4-Flash TP=4 could load and return `/model_info`, but long generation could fail when a running decode request was mixed with several new long-prefill requests. One observed failing shape was:

```text
#new-seq: 3
#new-token: 12288
#running-req: 1
```

The old workaround avoided this scheduler shape. The root-fix path should handle it without changing scheduler admission.

During validation, arbitrary flattened-row chunking was rejected: a forced 1024-row chunk test made `4096+16` generated token IDs diverge. The safe unit here is the request/prefill segment boundary from the scheduler.

## Validation

Validated on one GB200 node with `lmsysorg/sglang:deepseek-v4-grace-blackwell`, model `/data/models/DeepSeek-V4-Flash`, TP=4.

Server shape:

```bash
python3 -m sglang.launch_server \
  --model-path /data/models/DeepSeek-V4-Flash \
  --tp 4 \
  --moe-runner-backend flashinfer_mxfp4 \
  --disable-cuda-graph \
  --disable-flashinfer-autotune \
  --skip-server-warmup \
  --max-total-tokens 262144 \
  --max-running-requests 32 \
  --context-length 8192
```

Important: validation did **not** use `--max-prefill-tokens 8192`, did **not** use the scheduler guard workaround, and did **not** set `SGLANG_OPT_DG_PAGED_MQA_LOGITS_CHUNK_SIZE`.

Env used for stress validation:

```bash
SGLANG_DSV4_PREFILL_METADATA_CHUNK_SIZE=1024
SGLANG_JIT_DEEPGEMM_PRECOMPILE=0
SGLANG_OPT_USE_TILELANG_MHC_PRE=0
SGLANG_OPT_USE_TILELANG_MHC_POST=0
```

Tests:
- `PYTHONPATH=python pytest -q test/srt/test_deepseek_v4_flashmla_chunking.py`: `3 passed`
- full-model raw-ID requests, `input_len=4096`, `output_len=1500`:
  - `bs=4`: `ok_count=4/4`, elapsed `415.496s`, target output throughput `14.44 token/s`
  - `bs=8`: `ok_count=8/8`, elapsed `549.593s`, target output throughput `21.83 token/s`
  - `bs=16`: `ok_count=16/16`, elapsed `1031.461s`, target output throughput `23.27 token/s`

The server logs crossed the original failure family repeatedly, including:

```text
#new-seq: 3, #new-token: 12288, #running-req: 1
#new-seq: 3, #new-token: 12288, #running-req: 3
#new-seq: 3, #new-token: 12800, #running-req: 3
```

No `get_decoding_sched_meta`, `invalid argument`, `paged_mqa_logits_metadata`, `uses too much shared`, scheduler exception, traceback, or runtime error appeared in the validation grep.

### CODE REVIEW

File: python/sglang/srt/layers/attention/compressed/metadata.py [hunk 1]
```diff
@@ -139,13 +153,21 @@ def __post_init__(self):
...
+                    _c4,
+                    self.c4_page_size,
+                    deep_gemm.get_num_sms(),
+                )
 
-            if envs.SGLANG_OPT_DG_PAGED_MQA_LOGITS_CHUNK_SIZE.get() != -1:
+            if not isinstance(self.deep_gemm_metadata, torch.Tensor):
                 pass
             else:
                 assert isinstance(self.deep_gemm_metadata, torch.Tensor)
```
[the remark this answers is not in the document]
Comment (foraxe) [reply python/sglang/srt/layers/attention/compressed/metadata.py]: Fixed in c9e54e294. Removed the redundant pass/assert block; behavior is unchanged because both tensor and chunked metadata objects were already accepted by the call path. Verified with the focused GB200 pytest: PYTHONPATH=python pytest -q test/srt/test_deepseek_v4_flashmla_chunking.py -> 3 passed.

File: python/sglang/srt/layers/deep_gemm_wrapper/paged_mqa_logits.py [hunk 2]
```diff
@@ -22,24 +22,81 @@ def copy_(self, other: "_PagedMqaLogitsMetadata"):
...
+    seq_lens: torch.Tensor,
+    batch_size: int,
+    chunk_size: int,
+) -> List[Tuple[int, int]]:
+    lens = seq_lens.detach().flatten().to("cpu")
+    starts = [0]
+    for idx in range(1, batch_size):
+        if int(lens[idx].item()) < int(lens[idx - 1].item()):
+            starts.append(idx)
+    starts.append(batch_size)
```
[the remark this answers is not in the document]
Comment (foraxe) [reply python/sglang/srt/layers/deep_gemm_wrapper/paged_mqa_logits.py]: Fixed in c9e54e294 with the same tensorized boundary detection for paged-MQA metadata chunk ranges. Verified with the focused GB200 pytest: PYTHONPATH=python pytest -q test/srt/test_deepseek_v4_flashmla_chunking.py -> 3 passed.

File: python/sglang/srt/layers/attention/deepseek_v4_backend_radix.py [hunk 3]
```diff
@@ -125,6 +128,90 @@ def _create_dummy_paged_compress_data(compress_ratio: int):
...
+) -> List[Tuple[int, int]]:
+    if seq_lens is None or seq_lens.shape[0] != total_rows:
+        return [(0, total_rows)]
+
+    lens = seq_lens.detach().flatten().to("cpu")
+    starts = [0]
+    for idx in range(1, total_rows):
+        if int(lens[idx].item()) < int(lens[idx - 1].item()):
+            starts.append(idx)
+    starts.append(total_rows)
```
[the remark this answers is not in the document]
Comment (foraxe) [reply python/sglang/srt/layers/attention/deepseek_v4_backend_radix.py]: Fixed in c9e54e294. Boundary detection now uses tensor comparison plus torch.nonzero instead of per-row scalar .item() calls. Verified with the focused GB200 pytest: PYTHONPATH=python pytest -q test/srt/test_deepseek_v4_flashmla_chunking.py -> 3 passed.

**Comment by foraxe** [comment]
Hi, @fzyzcjy , could you please take a look?

**Comment by foraxe** [comment]
Manually closing this PR as it is no longer being pursued.
